zbus/src/address/mod.rs

```diff
@@ -87,6 +87,17 @@ impl Address {
     /// Get the address for the system bus respecting the `DBUS_SYSTEM_BUS_ADDRESS` environment
     /// variable. If we don't recognize the value (or it's not set) we fall back to
     /// `/var/run/dbus/system_bus_socket`.
+    ///
+    /// # Security
+    ///
+    /// <div class="warning">
+    ///
+    /// `zbus` does not perform any security checks or trust validation on the address (path)
+    /// returned by this method.
```

Comment:
sorry but I only noticed now: This method does not return an address. We need to make it more specific to the problematic bit: 

```suggestion
    /// Note that `zbus` does not (and can not) validate the value of the `DBUS_SYSTEM_BUS_ADDRESS`
    /// environment variable before using it. Users in security-sensitive contexts (ex: authentication)
    /// are responsible for ensuring that the value of this environment variable can be trusted.
```

Reply:
FWIW I don't agree with the exact wording here. I think its better for users to validate the value the function returns and not its inputs independently.

Reply:
Then we disagree on what's expected of the user. :laughing: Perhaps I'm missing something but:

* how does the user verify the connection is valid? I'd imagine it'd be easier to ensure that the env is correct/clean. I think this is what was suggested in [the summary of the CVE you pointed to](https://nvd.nist.gov/vuln/detail/CVE-2012-3524):
>  libdbus maintainers state that this is a vulnerability in the applications that do not cleanse environment variables
* the damage could already be done by the time user has a connection instance, since the connection creation involves a handshake and "Hello" message being sent out to the peer.

Reply:
@complexspaces are you convinced by my arguments? If not, could you please reply? I understand if you don't want to argue about this anymore but in that case, kindly accept my proposed wording so we can move on here. :pray:

Reply:
Hey again, sorry, I ended up at a conference that took most of my time and attention.

> how does the user verify the connection is valid? I'd imagine it'd be easier to ensure that the env is correct/clean. I think this is what was suggested in...

My argument here is that I'm not usually a fan of recommending users inspect all of the inputs to a function because its more fragile to drift over time or missing a possible branch. Instead, its easier to create an `Address::system` and analyze that as a "final piece" before creating a connection from it.

> the damage could already be done by the time user has a connection instance, since the connection creation involves a handshake and "Hello" message being sent out to the peer.

Even if you did `Connection::system` instead of `Address::system` I don't follow how a handshake "hello" message would be a security risk when using a memory-safe language like Rust.

Reply:
> Hey again, sorry, I ended up at a conference that took most of my time and attention.

No worries.

> > how does the user verify the connection is valid? I'd imagine it'd be easier to ensure that the env is correct/clean. I think this is what was suggested in...
> 
> My argument here is that I'm not usually a fan of recommending users inspect all of the inputs to a function because its more fragile to drift over time or missing a possible branch.

* Sorry but tbh that sounds a bit vague. How does that all translate to this specific case? :thinking: 
* If it makes you feel better, tehcnically there is no input to the function in question. This is about the environment that the function makes an assumption about.

> Instead, its easier to create an `Address::system` and analyze that as a "final piece" before creating a connection from it.

My main question here is how it is even possible to verify that the connection that you already created is pointing to what you intended it to point to? Perhaps it is very much possible and it's very easy (like you wrote here) but then you need to educate me a bit. :)

Reply:
> Sorry but tbh that sounds a bit vague. How does that all translate to this specific case? 🤔

Essentially I don't think its 100% reliable to just check the environment variable that `zbus` does because either:
- `zbus` _could_ add more logic in the `system` function that needs verified additionally by the end user.
- The default/fallback path should probably be verified as well for completeness.

> My main question here is how it is even possible to verify that the connection that you already created is pointing to what you intended it to point to?

Oh, I see what you are saying now, sorry. AFAIK you _need_ to go `Address` -> `ConnectionBuilder` manually to verify everything right. I could update the documentation to point at that path more strongly.

Reply:
> However, I'm still wondering if it's even possible to verify the address? If that is easily possible (I understood that's what you claimed in your previous comment?) then we should just do it ourselves, w/o even involving the user/programmer.

I mentioned this in my email previously, but there are some strategies 1P has adopted for path verification but its not anything confident enough in (when trying to analyze the entirety of how Linux filesystem permissions can work) that I would want to put it in `zbus` as a layer of protection for users. We are fine owning the risk as a company but I think its too whack-a-mole for FOSS 😅.

I think if people are doing security sensitive things with `zbus`, then they should be responsible for making this set of risk acceptances and functionality tradeoffs instead of making you responsible for upkeeping complicated security checks forever.

Reply:
> I think if people are doing security sensitive things with `zbus`, then they should be responsible for making this set of risk acceptances and functionality tradeoffs instead of making you responsible for upkeeping complicated security checks forever.

I agree but I feel like this is a lot more inline with my suggestion and the libdbus maintainer's conclusion. Having said that, I'm fine with telling people to verify the address (instead of ensuring their env is clean in general) if we can give them some clue how to achieve that (some link to documentation perhaps?).

Reply:
Do you have a rough idea what documentation you'd like to see there? I'm not opposed to your idea of doing both here.

Reply:
My idea was simply that we tell people to ensure their environment is clean and why so.